File: sidecar/app/routers/profiles.py
"""`/profiles` — voice library CRUD + lock/unlock + usage + audio.

Thin router over `services/profiles.py` (voice-profiles.md). Multipart for
create/lock (they carry a file or form id); JSON for the partial PUT.
"""

from fastapi import APIRouter, File, Form, UploadFile
from fastapi.responses import FileResponse
from pydantic import BaseModel

from ..services import profiles as svc

router = APIRouter(prefix="/profiles")


class ProfilePatch(BaseModel):
    name: str | None = None
    ref_text: str | None = None
    instruct: str | None = None
    language: str | None = None


@router.get("")
def list_profiles() -> list[dict]:
    return svc.list_profiles()


@router.post("")
async def create_profile(
    name: str = Form(...),
    ref_audio: UploadFile = File(...),
    ref_text: str = Form(""),
    instruct: str = Form(""),
    language: str = Form("Auto"),
    seed: int | None = Form(None),
) -> dict:
    audio_bytes = await ref_audio.read()
    return svc.create_profile(
        name=name,
        audio_bytes=audio_bytes,
        original_filename=ref_audio.filename,
        ref_text=ref_text,
        instruct=instruct,
        language=language,
        seed=seed,
    )


@router.get("/{profile_id}")
def get_profile(profile_id: str) -> dict:
    return svc.get_profile_or_404(profile_id)


@router.put("/{profile_id}")
def update_profile(profile_id: str, patch: ProfilePatch) -> dict:
    return svc.update_profile(
        profile_id,
        name=patch.name,
        ref_text=patch.ref_text,
        instruct=patch.instruct,
        language=patch.language,
    )


@router.get("/{profile_id}/audio")
def profile_audio(profile_id: str):
    path = svc.audio_path_for(profile_id)
    return FileResponse(str(path), media_type="audio/wav")


@router.get("/{profile_id}/usage")
def profile_usage(profile_id: str) -> dict:
    return svc.usage(profile_id)


@router.post("/{profile_id}/lock")
def lock_profile(
    profile_id: str,
    history_id: str = Form(...),
    seed: int | None = Form(None),
) -> dict:
    return svc.lock_profile(profile_id, history_id, seed)


@router.post("/{profile_id}/unlock")
def unlock_profile(profile_id: str) -> dict:
    return svc.unlock_profile(profile_id)


@router.delete("/{profile_id}")
def delete_profile(profile_id: str) -> dict:
    return svc.delete_profile(profile_id)

File: frontend/src/lib/api/profiles.ts
import { apiBase, apiDelete, apiFetch, apiJson, apiPost, apiPut } from "./client";
import type { ProfileUsage, VoiceProfile } from "./types";

export const listProfiles = () => apiJson<VoiceProfile[]>("/profiles");

export const getProfile = (id: string) => apiJson<VoiceProfile>(`/profiles/${id}`);

export interface CreateProfileInput {
  name: string;
  refAudio: Blob;
  refAudioFilename?: string;
  refText?: string;
  instruct?: string;
  language?: string;
  seed?: number | null;
}

export function createProfile(input: CreateProfileInput): Promise<{ id: string; name: string }> {
  const fd = new FormData();
  fd.set("name", input.name);
  fd.set("ref_audio", input.refAudio, input.refAudioFilename ?? "reference.wav");
  fd.set("ref_text", input.refText ?? "");
  fd.set("instruct", input.instruct ?? "");
  fd.set("language", input.language ?? "Auto");
  if (input.seed !== undefined && input.seed !== null) fd.set("seed", String(input.seed));
  return apiPost("/profiles", fd);
}

export interface ProfilePatch {
  name?: string;
  ref_text?: string;
  instruct?: string;
  language?: string;
}

export const updateProfile = (id: string, patch: ProfilePatch) =>
  apiPut<VoiceProfile>(`/profiles/${id}`, patch);

export const deleteProfile = (id: string) => apiDelete<{ deleted: string }>(`/profiles/${id}`);

export const getProfileUsage = (id: string) => apiJson<ProfileUsage>(`/profiles/${id}/usage`);

/** Full URL to a profile's representative audio (for an <audio> element). */
export const profileAudioUrl = (id: string) =>
  apiBase().then((base) => `${base}/profiles/${id}/audio`);

/** The profile's original reference clip as raw bytes (GET /profiles/{id}/audio),
 *  for downloading it to disk in its source format. Throws ApiError on failure. */
export async function profileAudioBytes(id: string): Promise<Uint8Array> {
  const res = await apiFetch(`/profiles/${id}/audio`);
  return new Uint8Array(await res.arrayBuffer());
}

export function lockProfile(id: string, historyId: string, seed?: number | null) {
  const fd = new FormData();
  fd.set("history_id", historyId);
  if (seed !== undefined && seed !== null) fd.set("seed", String(seed));
  return apiPost<{ locked: true; profile_id: string; locked_audio_path: string }>(
    `/profiles/${id}/lock`,
    fd,
  );
}

export const unlockProfile = (id: string) =>
  apiPost<{ unlocked: true; profile_id: string }>(`/profiles/${id}/unlock`);
